src/robot/running/bodyrunner.py

```diff
@@ -631,18 +638,18 @@ def _run_finally(self, data, run):
 
 class WhileLimit:
 
-    def __init__(self, on_limit_message=None):
+    def __init__(self, on_limit=None, on_limit_message=None):
+        self.on_limit = on_limit
         self.on_limit_message = on_limit_message
 
     @classmethod
-    def create(cls, limit, on_limit_message, variables):
+    def create(cls, limit, on_limit, on_limit_message, variables):
         if on_limit_message:
-            on_limit_message = variables.replace_string(
-                on_limit_message)
+            on_limit_message = variables.replace_string(on_limit_message)
```

Comment:
Possible error in replacing variables should probably be handled better and would need a test. Not really related to `on_limit` but could be fixed as part of this change anyway.